### ISSUE TITLE
Sign error in CNN Layer Visualization

### ISSUE THREAD

**Issue by McLawrence** [utkuozbulak/pytorch-cnn-visualizations#16]
There is only a very minor sign error in the code. In the hook version you correctly minimize the negative mean, hence, maximizing the mean.

In the unhooked version however, the minus sign is missing, computing the input than activates the feature the least (line 102):

`loss = torch.mean(self.conv_output)`

Additionally, the code has to be modified to run the latest pytorch 0.4, where zero-dimensional tensor can no longer be indexed (line 101 and 63).

**Comment by utkuozbulak** [comment]
Holy! I was sure I fixed that! I will look into it asap!

**Comment by herandy** [comment]
There is another problem that I have with version 0.4.0. Setting model.eval() my model doesn't learn anything and stays the same throughout all the epochs for both cnn visualization as well as generate class specific samples.

**Comment by utkuozbulak** [comment]
@herandy I think you are confusing concepts. model.eval() freezes the weights and converts the model into prediction mode. To train you need to call model.train(). 

If you have a dropout or batchnorm (or any custom layer that behaves that way) in the model and get visualizations while model is in train mode you will get a new visualization every time. 

For some of the discussion see:
https://stackoverflow.com/questions/48146926/whats-the-meaning-of-function-eval-in-torch-nn-module
https://github.com/pytorch/pytorch/issues/5406

**Comment by utkuozbulak** [comment]
It is to optimize the image, not the model. See what is being optimized:

`optimizer = Adam([self.processed_image], lr=0.1, weight_decay=1e-6)`

**Comment by herandy** [comment]
Thank you.

**Comment by utkuozbulak** [comment]
Fixed the error, thanks again for notifying!

**Comment by roachsinai** [comment]
Hi @utkuozbulak I don't understand the method you use to [Visualize Convolutional Neural Network Filter](https://github.com/utkuozbulak/pytorch-cnn-visualizations#convolutional-neural-network-filter-visualization)

> CNN filters can be visualized when we optimize the input image with respect to output of the specific convolution operation. 

So you mean as the network minimize the negetive mean of output, `self.created_image = recreate_image(self.processed_image)`, the `self.created_image` is the texture of the cnn filter? Dose the most corresponding image texture get the biggest activation is the reason?

**Comment by utkuozbulak** [comment]
The loss function aims to maximize the activation a specific filter with 
```python
loss = -torch.mean(self.conv_output)
```
and updates the input image in such a way that the average activation for specified conv output is (hopefully) higher at each step.

**Comment by roachsinai** [comment]
Yes, I know that. So at the end, the input image chaged to a image that maximize the activation of a specific filter?

**Comment by utkuozbulak** [comment]
Yes.
